sympy/core/basic.py

```diff
@@ -1525,6 +1525,57 @@ def match(self, pattern, old=False):
         pattern = sympify(pattern)
         return pattern.matches(self, old=old)
 
+    def _partition_args(self, var):
+        """
+        Return a `dict` of subexpressions which contains `var`.
+        Subexpressions which doesn't contain `var` are considered to be constant.
+
+        See:
+        https://people.eecs.berkeley.edu/~fateman/papers/partition-new2.pdf
+
+        Examples
+        ========
+
+        >>> from sympy import exp, sqrt, cos, sin
+        >>> from sympy.abc import x, a, b
+        >>> expr = a + b + x
+        >>> expr._partition_args(x)
+        {'Constant': [a, b], 'Symbol': [x]}
+        >>> expr = cos(a)*b*sin(x)
+        >>> expr._partition_args(x)
+        {'Constant': [b, cos(a)], 'sin': [sin(x)]}
+
+        When the expression head is not commutative('Add' and 'Mul' type
+        epressions), the same expression is returned in form of dict.
+
+        >>> exp(x)._partition_args(x)
+        {'Constant': [], 'exp': [exp(x)]}
+        >>> exp(a)._partition_args(x)
+        {'Constant': [exp(a)]}
+
+        """
+        from .add import Add
+        from .mul import Mul
+        leaves = {}
+        leaves['Constant'] = []
+
+        if not isinstance(self, (Mul, Add)):
+            if self.has(var):
```

Comment:
-> `if var in self.free_symbols`

(and the same change throughout because `has` does not imply dependence on a given symbol)

```
>>> Integral(x, (x, 1, 0)).has(x)
True
>>> Integral(x, (x, 1, 0)).free_symbols
{}
```